### PR TITLE
fix(network): PeerStore's cache gets out of sync

### PR DESCRIPTION
**Pull request by mooori**
`PeerStore` maintains its internal cache in two maps:
https://github.com/near/nearcore/blob/dddbb6a9691d0be59bef59cf50bbf096e1b617c8/chain/network/src/peer_manager/peer_store.rs#L47-L51

In [`remove_expired`](https://github.com/near/nearcore/blob/dddbb6a9691d0be59bef59cf50bbf096e1b617c8/chain/network/src/peer_manager/peer_store.rs#L282-L299) peers are deleted (via `delete_peers`) *only* from `peer_states`
but not from `addr_peers`.

Assume a peer with address `x` was deleted via `remove_expired`. When learning
about a new indirect peer which has the same address `x`, that information is
not stored due to `x` still being in `addr_peers`:
https://github.com/near/nearcore/blob/dddbb6a9691d0be59bef59cf50bbf096e1b617c8/chain/network/src/peer_manager/peer_store.rs#L381-L389

This PR updates `delete_peers` to remove addresses from `addr_peers` and
adds `test_delete_peers`.

There's only one other call site of `delete_peers`: `PeerStore::new`. Its
behavior is not changed by this PR. The peers it deletes aren't added to the
internal cache.

### CODE REVIEW

**Comment by mooori** [comment]
Prior to this change, the issue can be reproduced with the following test which
will fail. It's not included in the code as it may be confusing without the
context of this PR.

```rust
#[test]
fn check_cache_sync() {
    let store = create_test_store();
    let mut peer_store = PeerStore::new(store, &[], Default::default()).unwrap();

    let addr = get_addr(0);
    let peer_info = get_peer_info(get_peer_id("node".to_string()), Some(addr.clone()));
    peer_store.add_indirect_peers([peer_info].into_iter()).unwrap();

    // Make `remove_expired` delete the peer.
    let mut network_config = NetworkConfig::from_seed("42", 42);
    network_config.peer_expiration_duration = std::time::Duration::new(0, 0);
    peer_store.remove_expired(&network_config).unwrap();

    // Add a new peer which has the same address.
    let peer_info = get_peer_info(get_peer_id("new_node".to_string()), Some(addr.clone()));
    peer_store.add_indirect_peers([peer_info.clone()].into_iter()).unwrap();
    assert_peers_in_cache(&peer_store, &[peer_info.id], &[addr]);
}
```

File: chain/network/src/peer_manager/peer_store.rs [hunk 1]
```diff
@@ -792,6 +786,60 @@ mod test {
...
+        let expected: HashSet<PeerId> = HashSet::from_iter(expected.iter().cloned());
+        assert_eq!(stored_peers, expected);
+    }
+
+    fn assert_peers_in_cache(
+        peer_store: &PeerStore,
+        expected_peers: &[PeerId],
+        expected_addresses: &[SocketAddr],
+    ) {
+        let expected_peers: HashSet<&PeerId> = HashSet::from_iter(expected_peers.iter());
```
Comment (jakmeier) [inline chain/network/src/peer_manager/peer_store.rs:806]: Not important to change but here and a few lines below the `iter` call is unnecessary:
```suggestion
        let expected_peers: HashSet<&PeerId> = HashSet::from_iter(expected_peers);
```

**Review by jakmeier — APPROVED** [review]
LGTM, thanks for the PR. Very nice and clear description!

File: chain/network/src/peer_manager/peer_store.rs (hunk 1 above)
Comment (mooori) [reply chain/network/src/peer_manager/peer_store.rs]: Removed - I hope I haven't missed any unnecessary calls.

**Review by jakmeier — APPROVED** [review]

**Pull request merged**